dash-pipeline/bmv2/underlay.p4

```diff
@@ -0,0 +1,76 @@
+#include <core.p4>
+#include "dash_headers.p4"
+#include "dash_metadata.p4"
+
+control underlay(
+      inout headers_t hdr
+    , inout metadata_t meta
+#ifdef TARGET_BMV2_V1MODEL
+    , inout standard_metadata_t standard_metadata
+#endif // TARGET_BMV2_V1MODEL
+#ifdef TARGET_DPDK_PNA
+    , in    pna_main_input_metadata_t  istd
+#endif // TARGET_DPDK_PNA  
+    ) 
+{
+    action set_nhop(bit<9> next_hop_id) {
+#ifdef TARGET_BMV2_V1MODEL
+        standard_metadata.egress_spec = next_hop_id;
+#endif // TARGET_BMV2_V1MODEL
+
+#ifdef TARGET_DPDK_PNA
+#ifdef DPDK_PNA_SEND_TO_PORT_FIX_MERGED
+        // As of 2023-Jan-26, the version of the pna.p4 header file
+        // included with p4c defines send_to_port with a parameter
+        // that has no 'in' direction.  The following commit in the
+        // public pna repo fixes this, but this fix has not yet been
+        // copied into the p4c repo.
+        // https://github.com/p4lang/pna/commit/b9fdfb888e5385472c34ff773914c72b78b63058
+        // Until p4c is updated with this fix, the following line will
+        // give a compile-time error.
+        istd.input_port = next_hop_id;
```

Comment:
I believe assignments to istd.input_port are not permitted, because that `istd` struct is a direction `in` parameter (I think).

Have you tried compiling this code with p4c-dpdk open source compiler, with TARGET_DPDK_PNA and DPDK_PNA_SEND_TO_PORT_FIX_MERGED both #define'd?  I suspect the compiler would give a compile-time error for this line.

Not sure why you would want to assign to the input_port?  I would have expected instead that you would declare a user-defined metadata field to save the next_hop_id, and then use that user-defined metadata field in some later table lookup.

Reply:
Note: If the intent for PNA code is to send packets to the output port identified by `next_hop_id`, then you can delete the assignment `istd.input_port = next_hop_id`, and change the following line to `send_to_port(next_hop_id)`, and that will work pretty much the same as the corresponding BMv2 code above.

Reply:
```
root@dashdell:/home/amith/DASH-forked_main/DASH/dash-pipeline# make p4c-dpdk-pna
Compile P4 program bmv2/dash_pipeline.p4 for dpdk-pna ...
mkdir -p dpdk-pna/dash_pipeline.dpdk && \
chmod o+w dpdk-pna/dash_pipeline.dpdk && \
docker run \
--rm \
--name dash-p4c-dpdk-root \
-u dashuser \
-it \
-v /home/amith/DASH-forked_main/DASH/dash-pipeline/bmv2:/bmv2 \
-v /home/amith/DASH-forked_main/DASH/dash-pipeline/dpdk-pna:/dpdk-pna \
-w / \
sonicdash.azurecr.io/dash-p4c-dpdk:fcefd9cdc60  \
p4c-dpdk -DTARGET_DPDK_PNA -DPNA_CONNTRACK --pp dpdk-pna/dash_pipeline.dpdk/dash_pipeline.pp.p4 \
        -o dpdk-pna/dash_pipeline.dpdk/dash_pipeline.spec --arch pna \
        --bf-rt-schema dpdk-pna/dash_pipeline.dpdk/dash_pipeline.p4.bfrt.json \
        --p4runtime-files dpdk-pna/dash_pipeline.dpdk/dash_pipeline.p4.p4info.txt \
        bmv2/dash_pipeline.p4
bmv2/dash_outbound.p4(45): [--Wwarn=unused] warning: 'is_overlay_dip_v4_or_v6' is unused
    action route_service_tunnel(bit<1> is_overlay_dip_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_outbound.p4(47): [--Wwarn=unused] warning: 'is_overlay_dip_mask_v4_or_v6' is unused
                                bit<1> is_overlay_dip_mask_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_outbound.p4(49): [--Wwarn=unused] warning: 'is_overlay_sip_v4_or_v6' is unused
                                bit<1> is_overlay_sip_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_outbound.p4(51): [--Wwarn=unused] warning: 'is_overlay_sip_mask_v4_or_v6' is unused
                                bit<1> is_overlay_sip_mask_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_outbound.p4(53): [--Wwarn=unused] warning: 'is_underlay_dip_v4_or_v6' is unused
                                bit<1> is_underlay_dip_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_outbound.p4(55): [--Wwarn=unused] warning: 'is_underlay_sip_v4_or_v6' is unused
                                bit<1> is_underlay_sip_v4_or_v6,
                                       ^^^^^^^^^^^^^^^^^^^^^^^^
bmv2/underlay.p4(16): [--Wwarn=unused] warning: 'next_hop_id' is unused
    action set_nhop(bit<9> next_hop_id) {
                           ^^^^^^^^^^^
bmv2/dash_pipeline.p4(288): [--Wwarn=unused] warning: 'outbound_bytes_counter' is unused
            @Sai[type="sai_uint64_t", isreadonly="true"] bit<64> outbound_bytes_counter,
                                                                 ^^^^^^^^^^^^^^^^^^^^^^
bmv2/dash_pipeline.p4(289): [--Wwarn=unused] warning: 'inbound_bytes_counter' is unused
            @Sai[type="sai_uint64_t", isreadonly="true"] bit<64> inbound_bytes_counter,
                                                                 ^^^^^^^^^^^^^^^^^^^^^
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table outbound_ConntrackOut_conntrackOut. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table outbound_ConntrackIn_conntrackIn. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table outbound_outbound_ca_to_pa|dash_outbound_ca_to_pa. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table inbound_ConntrackIn_conntrackIn. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table inbound_ConntrackOut_conntrackOut. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table eni_meter. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table pa_validation. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table inbound_routing. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table meter_rule. Copying all match fields to metadata
[--Wwarn=mismatch] warning: Mismatched header/metadata struct for key elements in table meter_bucket. Copying all match fields to metadata
# p4c-dpdk -DDPDK_PNA -DPNA_CONNTRACK --dump dpdk-pna/dash_pipeline.dpdk --pp dpdk-pna/dash_pipeline.dpdk/dash_pipeline.pp.p4 \
# -o dpdk-pna/dash_pipeline.dpdk/dash_pipeline.spec --arch pna --bf-rt-schema dpdk-pna/dash_pipeline.dpdk/dash_pipeline.p4.bfrt.json \
# --p4runtime-files dpdk-pna/dash_pipeline.dpdk/dash_pipeline.p4.p4info.txt bmv2/dash_pipeline.p4
root@dashdell:/home/amith/DASH-forked_main/DASH/dash-pipeline# 
```

The compiler doesn't throw any error.

Reply:
Your compilation run did NOT define the symbol DPDK_PNA_SEND_TO_PORT_FIX_MERGED, so the offending code was #ifdef'd out and thus not visible to the compiler, hence no error.

I see you made another commit to make my suggested change, which is definitely better, or at the very least, the v1model and PNA versions specify similar behavior.